experimental/Schemes/src/ToricBlowups/methods.jl

```diff
@@ -1,3 +1,231 @@
+  S = cox_ring(domain(f))
+  new_var = S[index_of_new_ray(f)]
+  J = total_transform(f, I)
+  return saturation_with_index(J, ideal(new_var))
+end
+
+@doc raw"""
+    cox_ring_group_homomorphism(f::ToricBlowupMorphism, g::MPolyDecRingElem) -> MPolyDecRingElem
+    cox_ring_group_homomorphism(f::ToricBlowupMorphism, I::MPolyIdeal) -> MPolyIdeal
+
+Let $f\colon Y \to X$ be the toric blowup corresponding to a star subdivision along a ray. Let $R$ and $S$ be the Cox rings of $X$ and $Y$, respectively.
+Considering $R$ and $S$ with their abelian group structures, there is a group homomorphism $\Phi\colon R \to S$ sending homogeneous ideals to homogeneous ideals with the defining property that the the total transform of a subscheme of $X$, given by a homogeneous ideal $I$ in $R$, is the subscheme of $Y$ given by the homogeneous ideal $\Phi(I)$ in $S$.
+
+This is implemented under the following assumptions:
+  * the star subdivision adds a ray (instead of subdividing along an existing ray),
+  * $X$ is an orbifold (meaning its fan is simplicial), and
+  * $X$ has no torus factors (meaning the rays span $N_{\mathbb{R}}$).
+
+# Examples
+```jldoctest
+julia> X = affine_space(NormalToricVariety, 2)
+Normal toric variety
+
+julia> f = blow_up(X, [2, 3])
+Toric blowup morphism
+
+julia> R = cox_ring(X)
+Multivariate polynomial ring in 2 variables over QQ graded by
+  x1 -> 1 by 0 empty matrix
+  x2 -> 1 by 0 empty matrix
+
+julia> (x1, x2) = gens(R)
+2-element Vector{MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}}:
+ x1
+ x2
+
+julia> cox_ring_group_homomorphism(f, x1 + x2)
+x1*e^2 + x2*e^3
+```
+"""
+cox_ring_group_homomorphism
+
+function cox_ring_group_homomorphism(f::ToricBlowupMorphism, g::MPolyDecRingElem)
+  @req parent(g) === cox_ring(codomain(f)) "g must be an element of the Cox ring of the codomain of f"
+  d, phi_exponents = _cox_ring_group_homomorphism_data(f)
+  S = cox_ring(domain(f))
+  new_var = S[index_of_new_ray(f)]
+
+  # Assuming the i-th variable of `cox_ring(X)` is the i-th variable of `cox_ring(Y)`
+  S_vars = elem_type(S)[S[i] for i in 1:nvars(S)]
+
+  make_S_term(c, exps) = c*prod(map(^, S_vars, exps))
+  coeff_exps = collect(zip(coefficients(g), exponents(g)))
+  h = S(0)
+  for (c, exps) in coeff_exps
+    new_exp = cdiv(sum(phi_exponents.*exps), d)
+    h += new_var^new_exp * make_S_term(c, exps)
+  end
+  return h
+end
+
+function cox_ring_group_homomorphism(f::ToricBlowupMorphism, I::MPolyIdeal)
+  S = cox_ring(domain(f))
+  return ideal(S, [cox_ring_group_homomorphism(f, g) for g in gens(I)])
+end
+
+@attr Tuple{ZZRingElem, Vector{ZZRingElem}} function _cox_ring_group_homomorphism_data(f::ToricBlowupMorphism)
+  # f: Y -> X
+  X = codomain(f)
+  Y = domain(f)
+  @req is_normal(X) "Only implemented when the variety is normal"
+  @req !has_torusfactor(X) "Only implemented when there are no torus factors"
+  @req n_rays(Y) == n_rays(X) + 1 "Only implemented when the blowup adds a ray"
```

Comment:
Referring to one of my above comments: It seems that here the sanity of the input is checked. Maybe these checks are then also triggered by the above code. Either way: It looks like some of these tests could be rather expensive, so I would like to point out the option to use `@check` instead of `@req` here so that turning of these tests internally is possible. What do you think?

Reply:
https://github.com/oscar-system/Oscar.jl/pull/4154#discussion_r1884136347